graphrag/index/flows/create_final_community_reports.py

```diff
@@ -0,0 +1,167 @@
+    nodes_input: pd.DataFrame,
+    edges_input: pd.DataFrame,
+    claims_input: pd.DataFrame | None,
+    callbacks: VerbCallbacks,
+    cache: PipelineCache,
+    strategy: dict,
+    async_mode: AsyncType = AsyncType.AsyncIO,
+    num_threads: int = 4,
+    covariates_enabled: bool = False,
+    full_content_text_embed: dict | None = None,
+    summary_text_embed: dict | None = None,
+    title_text_embed: dict | None = None,
+) -> pd.DataFrame:
+    """All the steps to transform community reports."""
+    nodes = _prep_nodes(nodes_input)
+    edges = _prep_edges(edges_input)
+
+    claims = None
+    if covariates_enabled and claims_input is not None:
+        claims = _prep_claims(claims_input)
+
+    community_hierarchy = restore_community_hierarchy_df(nodes)
+
+    local_contexts = prepare_community_reports_df(
+        nodes, edges, claims, callbacks, strategy.get("max_input_length", 16_000)
+    )
+
+    community_reports = await create_community_reports_df(
+        local_contexts,
+        nodes,
+        community_hierarchy,
+        callbacks,
+        cache,
+        strategy,
+        async_mode=async_mode,
+        num_threads=num_threads,
+    )
+
+    community_reports["id"] = community_reports["community"].apply(
+        lambda _x: str(uuid4())
+    )
+
+    # Embed full content if not skipped
+    if full_content_text_embed:
+        community_reports["full_content_embedding"] = await text_embed_df(
+            community_reports,
+            callbacks,
+            cache,
+            column="full_content",
+            strategy=full_content_text_embed["strategy"],
+            embedding_name="community_report_full_content",
+        )
+
+    # Embed summary if not skipped
+    if summary_text_embed:
+        community_reports["summary_embedding"] = await text_embed_df(
+            community_reports,
+            callbacks,
+            cache,
+            column="summary",
+            strategy=summary_text_embed["strategy"],
+            embedding_name="community_report_summary",
+        )
+
+    # Embed title if not skipped
+    if title_text_embed:
+        community_reports["title_embedding"] = await text_embed_df(
+            community_reports,
+            callbacks,
+            cache,
+            column="title",
+            strategy=title_text_embed["strategy"],
+            embedding_name="community_report_title",
+        )
+
+    return community_reports
+
+
+def _prep_nodes(input: pd.DataFrame) -> pd.DataFrame:
+    input = input.fillna(value={NODE_DESCRIPTION: "No Description"})
+    # merge values of four columns into a map column
+    input[NODE_DETAILS] = input.apply(
+        lambda x: {
+            NODE_ID: x[NODE_ID],
+            NODE_NAME: x[NODE_NAME],
+            NODE_DESCRIPTION: x[NODE_DESCRIPTION],
+            NODE_DEGREE: x[NODE_DEGREE],
+        },
+        axis=1,
+    )
+    return input
+
+
+def _prep_edges(input: pd.DataFrame) -> pd.DataFrame:
```

Comment:
Idem. Rule of thumb: Only use apply when there's no other way

Reply:
Andres also seemed to think these could be simplified with a to_dict, but it didn't work on first try so needs a little more investigation